atc/worker/client.go

```diff
@@ -285,6 +316,52 @@ func (client *client) chooseTaskWorker(
 				}
 				return nil, err
 			}
+
+			if !existingContainer {
+				// Enter the queue and get the position, or just get the position if already in queue
+				taskId := fmt.Sprintf("%d_%d_%d", metadata.PipelineID, metadata.JobID, metadata.BuildID)
+				if dedicatedWorker {
+					teamId = workerSpec.TeamID
+				} else {
+					teamId = 0
+				}
+				tags := containerSpec.Tags
+				sort.Strings(tags)
+				workerTags := strings.Join(tags, "_")
+				queue_position, qlen, err := client.tasksQueue.FindOrAppend(taskId, workerSpec.Platform, teamId, workerTags, logger)
+				defer client.tasksQueue.Leave(taskId, logger)
```

Comment:
1. I want to keep the full logic in the TasksQueue and not in Client. This is similar, for instance, to `pool.FindOrChooseWorkerForContainer`
2. It's called multiple times and it's idempotent. This saves the complexity that would be required just to know if the task is being queued on this cycle or it was before.

Reply:
>  2\. It's called multiple times and it's idempotent. This saves the complexity that would be required just to know if the task is being queued on this cycle or it was before.

It might be idempotent from the point of view of the side-effects on the database.

The problem is on the Go runtime. What I think will happen is that the `defer`s will add up to the runtime defer linked list, each time we enter this code path. Depending on how often this code path is taken for the same queue element, this can have a serious impact on the runtime resources and behavior.

Reply:
After giving some thought I implemented an easy solution to avoid calling the `defer` more than once.
Thank you for the feedback.